### PR TITLE
run drm model and visualize result

### PR DESCRIPTION
**Pull request by JEN6YT**
Pytorch implementation of drm + Visualization with 10 training epochs

### CODE REVIEW

**Comment by JEN6YT** [comment]
percentile_barrier Implementation:

- wasn't able to test it on local or aws vm

Screenshot 2024-12-03 at 8 47 46 PM

File: run_script.py [hunk 1]
```diff
@@ -51,10 +60,84 @@
...
 
 
+# ----- DRM ----- # 
+
+from Model.drm import *
+
+drm_model = SimpleTCModelDNN(input_dim= 46, num_hidden= 92)
+
+# Split data into treated and untreated
+train_treat_index = np.where(w_tr==1)[0]
```
Comment (wzoustanford) [inline run_script.py:69]: This looks good. 

For good coding practice, I wonder if we can make the code shorter? 

1. could we just do e.g. nX_tr[np.where(w_tr==1)[0]]? this saves re-typing the variable 'train_treat_index' 
2. I really don't know if this is true, but have you thought about: is it easier to do indexing After converting to torch.tensor? torch may have operators (e.g. torch.where torch.index_select) that could make the notation shorter. (by indexing I mean selecting parts of the tensor with a binary mask e.g. nX_tr[train_treat_index])

File: Model/drm.py [hunk 2]
```diff
@@ -1,8 +1,121 @@
...
+        
+        return h_tre_rnkscore, h_unt_rnkscore
+
+
+def compute_objective(h_tre_rnkscore, h_unt_rnkscore, c_tre, c_unt, o_tre, o_unt):
+    """
+    Computes the cost-gain effectiveness objective.
+
+    Args:
+    - h_tre_rnkscore: torch.Tensor, ranker scores for treatment group.
```
Comment (wzoustanford) [inline Model/drm.py:53]: love the comments here as documentation, it is very good practice.

File: Model/drm.py [hunk 3]
```diff
@@ -1,8 +1,121 @@
...
+    - h_unt_rnkscore: torch.Tensor, ranker scores for control group.
+    - c_tre: torch.Tensor, cost labels for treatment group.
+    - c_unt: torch.Tensor, cost labels for control group.
+    - o_tre: torch.Tensor, order labels for treatment group.
+    - o_unt: torch.Tensor, order labels for control group.
+
+    Returns:
+    - obj: torch.Tensor, the computed objective.
+    """
+    s_tre = F.softmax(h_tre_rnkscore.squeeze(), dim=0)
```
Comment (wzoustanford) [inline Model/drm.py:63]: [minor] squeeze removes the dim 1, but wonder where the dimension 1 is from, could we remove it upstream to make the code cleaner.

File: Model/drm.py [hunk 4]
```diff
@@ -1,8 +1,121 @@
...
+    s_unt = F.softmax(h_unt_rnkscore.squeeze(), dim=0)
+
+    
+    dc_tre = torch.sum(s_tre * c_tre)
```
Comment (wzoustanford) [inline Model/drm.py:67]: [minor] perhaps add a 'dim=x' argument to the sum function, makes the tensor dimensions more readable.

File: Model/drm.py [hunk 5]
```diff
@@ -1,8 +1,121 @@
...
+    dc_unt = torch.sum(s_unt * c_unt)
+
+    do_tre = torch.sum(s_tre * o_tre)
+    do_unt = torch.sum(s_unt * o_unt)
+
+    obj = (dc_tre - dc_unt) / (do_tre - do_unt)
```
Comment (wzoustanford) [inline Model/drm.py]: experiment with this first and see if we obtain the same graph: but should we add the 'relu' rectifier both the denominator and numerator? This ensures the operators are continuous, and the objective function is differentiable w.r.t. the parameters. (1/x is not differentiable at point x = 0)

File: Model/drm.py [hunk 6]
```diff
@@ -1,8 +1,121 @@
...
+
+    # Optional differentiable version:
```
Comment (wzoustanford) [inline Model/drm.py:75]: it's here, supposed this line was tried in the experiment?

File: Model/drm.py [hunk 7]
```diff
@@ -1,8 +1,121 @@
...
+    - c_unt: torch.Tensor, cost labels for control group.
+    - o_tre: torch.Tensor, order labels for treatment group.
+    - o_unt: torch.Tensor, order labels for control group.
+    - lr: float, learning rate.
+    - save_path: path, path to save checkpoint
+
+    Returns:
+    - obj: torch.Tensor, the computed objective.
+    """
+    optimizer = Adam(model.parameters(), lr=lr)
```
Comment (wzoustanford) [inline Model/drm.py:97]: Optimizer should also be defined only once, then used at each iteration. 

For code readability, I'd suggest remove the 'optimize_model' function, since it confuses readers to consider initializing the optimizer inside it due to its name, also you'd have to document its argument while it only wraps 3-5 lines.

File: Model/drm.py [hunk 8]
```diff
@@ -1,8 +1,121 @@
...
+
+    h_tre_rnkscore, h_unt_rnkscore = model(D_tre, D_unt)
+    obj = compute_objective(h_tre_rnkscore, h_unt_rnkscore, c_tre, c_unt, o_tre, o_unt)
+
+    optimizer.zero_grad()
+    (-obj).backward()  # Negative objective for maximization
+    optimizer.step()
```
Comment (wzoustanford) [inline Model/drm.py]: this line, if I understand correctly, only takes 1 gradient step.

File: run_script.py [hunk 9]
```diff
@@ -51,10 +60,84 @@
...
+treat_value_va = torch.tensor(treat_value_va, dtype=torch.float32)
+untreat_value_va = torch.tensor(untreat_value_va, dtype=torch.float32)
+
+h_tre_rnkscore, h_unt_rnkscore = drm_model.forward(D_tre=treat_nX_tr, D_unt=untreat_nX_tr)
+
+# Training
+drm_epochs = 10
+save_path="model.pth"
+
+for epoch in range(drm_epochs):
```
Comment (wzoustanford) [inline run_script.py]: if I understand correctly, the code here only implements 10 gradient steps of Adam. 

- let's run it for longer, print out the loss to ensure it converges (run to the point where loss value change is small across steps) 
- only for clarification, 'epoch' is commonly used in mini-batch or stochastic gradient descent, to indicate running enough data points to cover the entire dataset. It's not ill-used here, since each step uses the entire dataset, it's just more commonly used in mini-batch settings. 'for step in range(drm_steps)' would also be OK here.

**Review by wzoustanford — CHANGES_REQUESTED** [review]
read and pls address the comments inline :)

File: Model/drm.py (hunk 7 above)
Comment (JEN6YT) [reply Model/drm.py:97]: Optimizer initialization.
Zero-grad

for
   obj
   backward
   step

**Comment by wzoustanford** [comment]
Has this problem been fixed? I'm guessing it is, since we already saw curves from the PB method. If so, feel free to reply to or close the comment. 

> percentile_barrier Implementation:
> 
> * wasn't able to test it on local or aws vm
> 
> Screenshot 2024-12-03 at 8 47 46 PM

File: run_script.py (hunk 9 above)
Comment (wzoustanford) [reply run_script.py]: @JEN6YT has this problem (only 10 gradient steps, and grad is zeroed every iteration) fixed? Is this same code logic used to run the two latest experiment curves. 

It's important to fix this optimization problem before comparing results. 

It's going to affect DRM and PB methods a lot.

File: run_script.py (hunk 1 above)
Comment (wzoustanford) [reply run_script.py:69]: Code quality is as important as getting results :) pls address this. 

You'll see that once the code base gets larger, the code and parameters are going to be messy. Engineering practices to make it readable and systematic, come first.

File: run_script.py (hunk 9 above)
Comment (wzoustanford) [reply run_script.py]: the fix is in this commit: https://github.com/JEN6YT/HCL/pull/3/commits/a62569fbf9878d87779e91cd761485512d1d975c

File: Model/drm.py (hunk 5 above)
Comment (wzoustanford) [reply Model/drm.py]: fixed in commit: https://github.com/JEN6YT/HCL/pull/3/commits/a62569fbf9878d87779e91cd761485512d1d975c
